helix-core/src/main/java/org/apache/helix/controller/stages/TopStateHandoffReportStage.java

```diff
@@ -469,16 +475,17 @@ private void reportTopStateComesBack(ResourceControllerDataProvider cache, Map<S
       }
     }
 
+    long duration = handOffEndTime - handOffStartTime;
+    long helixLatency = duration - fromTopStateUserLatency - toTopStateUserLatency;
+    // It is possible that during controller leader switch, we lost previous master information
+    // and use current time to approximate missing top state start time. If we see the actual
+    // user latency is larger than the duration we estimated, we use user latency to start with
+    duration = Math.max(duration, helixLatency);
+    boolean isGraceful = record.isGracefulHandoff();
+    logMissingTopStateInfo(duration, helixLatency, isGraceful, partition.getPartitionName());
+    // TODO: Following threshold check can be removed and stats can be reported even if hand-off took more than
+    //  threshold.
     if (handOffStartTime > 0 && handOffEndTime - handOffStartTime <= threshold) {
```

Comment:
i don't understand why are you moving this logic above?

Reply:
Currently in this code path (when top state is recovered but handoff happened beyond threshold) we don't update the metric hence we don't log as well. But this makes it difficult to debug if we want to see what the exact handoff duration in this situation. Hence I moved just the logging upward but kept metric reporting as it is.

Reply:
one more thing, metric log is exactly same as what Jacob wants, correct? i don't see resource name printed and so kind of curious that just having partition name enough?

Reply:
Yes I didn’t make any changes in logs so it should be same as what they were using before. I made minor changes in log statement in above code path as we don’t have any handoff duration to report(because handoff has not completed yet)

Reply:
It may be the case but that's existing behavior (with assumption mostly it will successful handoff meaning handoff happening within threshold)  and asked by espresso folks to have their tools/debugging easier.